sdk/cosmosdb/cosmos/src/globalPartitionEndpointManager.ts

```diff
@@ -0,0 +1,382 @@
+        const partitionFailOver =
+          this.partitionKeyRangeToLocationForReadAndWrite.get(partitionKeyRangeId);
+
+        const canCircuitBreakerTriggerPartitionFailOver =
+          await partitionFailOver.canCircuitBreakerTriggerPartitionFailOver(
+            isReadRequest(requestContext.operationType),
+          );
+        if (canCircuitBreakerTriggerPartitionFailOver) {
+          return {
+            overridden: true,
+            newLocation: partitionFailOver.currentEndPoint,
+          };
+        }
+      }
+    }
+    return { overridden: false };
+  }
+
+  /**
+   * Increments the failure counter for the specified partition and checks if the partition can fail over.
+   * This method is used to determine if a partition should be failed over based on the number of request failures.
+   */
+  private async incrementFailureCounterAndCheckFailover(
+    requestContext: RequestContext,
+  ): Promise<boolean> {
+    const partitionKeyRangeId = requestContext.partitionKeyRangeId;
+    const failedEndPoint = requestContext.endpoint;
+    let partitionKeyRangeFailoverInfo: PartitionKeyRangeFailoverInfo;
+
+    if (this.isRequestEligibleForPerPartitionAutomaticFailover(requestContext)) {
+      if (!this.partitionKeyRangeToLocationForWrite.has(partitionKeyRangeId)) {
+        // If the partition key range is not already in the map, add it
+        const failoverInfo = new PartitionKeyRangeFailoverInfo(failedEndPoint);
+        this.partitionKeyRangeToLocationForWrite.set(partitionKeyRangeId, failoverInfo);
+      }
+      partitionKeyRangeFailoverInfo =
+        this.partitionKeyRangeToLocationForWrite.get(partitionKeyRangeId);
+    } else if (this.isRequestEligibleForPartitionLevelCircuitBreaker(requestContext)) {
+      if (!this.partitionKeyRangeToLocationForReadAndWrite.has(partitionKeyRangeId)) {
+        // If the partition key range is not already in the map, add it
+        const failoverInfo = new PartitionKeyRangeFailoverInfo(failedEndPoint);
+        this.partitionKeyRangeToLocationForReadAndWrite.set(partitionKeyRangeId, failoverInfo);
+      }
+      partitionKeyRangeFailoverInfo =
+        this.partitionKeyRangeToLocationForReadAndWrite.get(partitionKeyRangeId);
+    }
+
+    if (!partitionKeyRangeFailoverInfo) {
+      return false;
+    }
+
+    const currentTimeInMilliseconds = Date.now();
+    await partitionKeyRangeFailoverInfo.incrementRequestFailureCounts(
+      isReadRequest(requestContext.operationType),
+      currentTimeInMilliseconds,
+    );
+
+    return partitionKeyRangeFailoverInfo.canCircuitBreakerTriggerPartitionFailOver(
+      isReadRequest(requestContext.operationType),
+    );
+  }
+
+  /** Validates if the given request is eligible for partition failover. */
+  private async isRequestEligibleForPartitionFailover(
+    requestContext: RequestContext,
+    shouldValidateFailedLocation: boolean,
+  ): Promise<boolean> {
+    if (
+      !requestContext ||
```

Comment:
Can we reverse this check? requestContext && requestContext.operationType ....
It would be simple and better for Non PPAF scenario.